### PR TITLE
Master

### PR DESCRIPTION
**Pull request by GiuseppeFasanella**
Replace GetCorrTree of EnergyScaleCorrection_class 
Small remaining issue: it doesn't work if the file is on the /eos/project  directory

### CODE REVIEW

File: ZFitter/bin/ZFitter.cpp [hunk 1]
```diff
@@ -160,13 +160,17 @@ void UpdateFriends(tag_chain_map_t& tagChainMap, TString regionsFileNameTag)
...
 					          << " to tag " << tag_chain_itr->first << std::endl;
-					chain->AddFriend(chain_itr->second.get());
+					  chain->AddFriend(chain_itr->second.get());
 				} // already added
 			}
 
-			if(chain->GetEntries() != chain_itr->second->GetEntries()) {
-				std::cerr << "[ERROR] Not the same number of events: " << chain->GetEntries() << "\t" << chain_itr->second->GetEntries() << std::endl;
-				exit(1);
+			if(chain->GetEntries() != chain_itr->second->GetEntries() && chain_itr->first!="scaleEle") {//Not scaleEle: check data and MC
```
Comment (shervin86) [inline ZFitter/bin/ZFitter.cpp]: why you don't want to check the number of events in the tree?

File: ZFitter/bin/ZFitter.cpp [hunk 2]
```diff
@@ -830,13 +834,39 @@ int main(int argc, char **argv)
...
 				exit(1);
 			}
-#ifdef toBeFixed
-			TTree *corrTree = eScaler.GetCorrTree(ch, "runNumber", "R9Eleprime");
-#else
-			TTree *corrTree = NULL;
-#endif
-			corrTree->SetName(TString("scaleEle_") + corrEleType.c_str());
-			corrTree->SetTitle(corrEleType.c_str());
+
```
Comment (shervin86) [inline ZFitter/bin/ZFitter.cpp]: No! This sould be done by addBranch, this should be completely removed

File: ZFitter/src/addBranch_class.cc [hunk 3]
```diff
@@ -515,7 +515,61 @@ TTree* addBranch_class::AddBranch_smearerCat(TChain* originalChain, TString tree
...
+  //To do:
+  //Add flexibility in .dat file name
+  //Add flexibility in branch names
+  //how to change doScale BEFORE creating escaler (otherwise default is false and if doScale is false correction file is not parsed)-->can't we just change doScale=true by default
+  EnergyScaleCorrection_class eScaler("Calibration/ZFitter/ScalesSmearings/Winter_2016_reReco_v1_ele",0,true,false);
+  originalChain->SetBranchStatus("*",0);
+  originalChain->SetBranchStatus("runNumber",1);
+  originalChain->SetBranchStatus("etaSCEle",1);
+  originalChain->SetBranchStatus("R9Ele",1);
+  originalChain->SetBranchStatus("energy_ECAL_ele",1);
```
Comment (shervin86) [inline ZFitter/src/addBranch_class.cc]: you need energyBranchName to be a string given to addBranch_class in the constructor or using a method setting a private member. Don't need to active all the energy branches

File: ZFitter/src/addBranch_class.cc [hunk 4]
```diff
@@ -515,7 +515,61 @@ TTree* addBranch_class::AddBranch_smearerCat(TChain* originalChain, TString tree
...
+  originalChain->SetBranchAddress("R9Ele", R9Ele);
+  //originalChain->SetBranchAddress("energy_ECAL_ele", energy_ECAL_ele);
+  originalChain->SetBranchAddress("energy_ECAL_pho", energy_ECAL_pho);
+  originalChain->SetBranchAddress("energySCEle_corr", energy_ECAL_ele);
+
+  Float_t scaleEle_[3];
+  TTree *newTree = new TTree(treename,""); //treeName is scaleEle
+  newTree->Branch("scaleEle", scaleEle_, "scaleEle[3]/F");
+  if(isMC){//dummy scale corr for MC -->probably there is a better way
+    scaleEle_[0]=-999;
```
Comment (shervin86) [inline ZFitter/src/addBranch_class.cc]: This has no effect, the tree is not filled

File: ZFitter/src/addBranch_class.cc [hunk 5]
```diff
@@ -515,7 +515,61 @@ TTree* addBranch_class::AddBranch_smearerCat(TChain* originalChain, TString tree
...
+    scaleEle_[1]=-999;
+    return newTree;
+  }
+
+  Long64_t nentries = originalChain->GetEntries();
+  //for(Long64_t ientry = 0; ientry < nentries; ientry++) {
+  for(Long64_t ientry = 0; ientry < 1000; ientry++) {
```
Comment (shervin86) [inline ZFitter/src/addBranch_class.cc]: nentries, not 1000

File: ZFitter/src/auxFunctions.cc [hunk 6]
```diff
@@ -9,6 +9,8 @@ std::string energyBranchNameFromInvMassName(std::string invMass_var)
 	else if(invMass_var == "invMass_5x5SC") energyBranchName = "energy_5x5SC";
 	else if(invMass_var == "invMass") energyBranchName = "energyEle";
 	else if(invMass_var == "invMass_rawSC") energyBranchName = "rawEnergySCEle";
+	else if(invMass_var == "invMass_SC_corr") energyBranchName = "energySCEle_corr";
```
Comment (shervin86) [inline ZFitter/src/auxFunctions.cc]: These two branches do not exist anymore

File: ZFitter/bin/ZFitter.cpp (hunk 1 above)
Comment (GiuseppeFasanella) [inline ZFitter/bin/ZFitter.cpp]: Normally addBranch creates friend root trees for data and MC and then checks the number of events. I want this check to be done only on data if the friend branch is "scaleEle" (no scaleEle branch for MC)

File: ZFitter/src/addBranch_class.cc (hunk 4 above)
Comment (GiuseppeFasanella) [inline ZFitter/src/addBranch_class.cc]: Why not?

File: ZFitter/src/auxFunctions.cc (hunk 6 above)
Comment (GiuseppeFasanella) [inline ZFitter/src/auxFunctions.cc]: They won't exist anymore in the next production, they still exist in the latest ntuples from miniAOD. I suggest to remove these two branches later on

**Comment by shervin86** [comment]
I would avoid many many small classes doing almost similar things.
The new tree with scaleEle and smearEle should be done by addBranch_class.
The necessary changes has to be done

**Pull request closed without merging**